crypto/evp/kdf_lib.c

```diff
@@ -0,0 +1,176 @@
+/*
+ * Copyright 2018 The OpenSSL Project Authors. All Rights Reserved.
+ * Copyright (c) 2018, Oracle and/or its affiliates.  All rights reserved.
+ *
+ * Licensed under the OpenSSL license (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#include <stdio.h>
+#include <stdlib.h>
+#include "internal/cryptlib.h"
+#include <openssl/engine.h>
+#include <openssl/evp.h>
+#include <openssl/x509v3.h>
+#include <openssl/kdf.h>
+#include <openssl/kdferr.h>
+#include "internal/asn1_int.h"
+#include "internal/evp_int.h"
+#include "internal/numbers.h"
+
+typedef int sk_cmp_fn_type(const char *const *a, const char *const *b);
+
+/* This array needs to be in order of NIDs */
+static const EVP_KDF_METHOD *standard_methods[] = {
+    &pbkdf2_kdf_meth,
+#ifndef OPENSSL_NO_SCRYPT
+    &scrypt_kdf_meth,
+#endif
+    &tls1_prf_kdf_meth,
+    &hkdf_kdf_meth,
+};
+
+DECLARE_OBJ_BSEARCH_CMP_FN(const EVP_KDF_METHOD *, const EVP_KDF_METHOD *,
+                           kmeth);
+
+static int kmeth_cmp(const EVP_KDF_METHOD *const *a,
+                     const EVP_KDF_METHOD *const *b)
+{
+    return ((*a)->nid - (*b)->nid);
+}
+
+IMPLEMENT_OBJ_BSEARCH_CMP_FN(const EVP_KDF_METHOD *, const EVP_KDF_METHOD *,
+                             kmeth);
+
+static const EVP_KDF_METHOD *kdf_meth_find(int type)
+{
+    EVP_KDF_METHOD tmp;
+    const EVP_KDF_METHOD *t = &tmp, **ret;
+
+    tmp.nid = type;
+    ret = OBJ_bsearch_kmeth(&t, standard_methods,
+                            OSSL_NELEM(standard_methods));
+    if (ret == NULL || *ret == NULL)
+        return NULL;
+    return *ret;
+}
+
+EVP_KDF_CTX *EVP_KDF_CTX_new_id(int id)
```

Comment:
Other EVP CTX functions are called just EVP_XYZ_CTX_new() why this one is _new_id() instead ?
I found it odd as I was trying to call EVP_KDF_CTX_new() and found it didn't exist...

Reply:
It takes a NID as a parameter and is modeled after the `EVP_PKEY_CTX_new_id(int id, ENGINE *e)` function, which is used for the existing PKEY KDF support, except that `EVP_KDF_CTX_new_id(int id)` doesn't take an `ENGINE*` parameter. Engine support is likely to be replaced with something else at some point in the future.

Many of the other `*_CTX_new()` functions take no parameters and the resulting context is not associated with any particular algorithm (i.e. NID).

Another question that could legitimately be asked is why the Derive function is called `EVP_KDF_derive()` instead of `EVP_KDF_CTX_derive()` or even `EVP_Derive()` (like `EVP_EncryptFinal()`). Well, the answer is that it is modeled after the `EVP_PKEY_derive()` function, which is used for the existing PKEY KDF support. I am not particularly attached to the current names and am happy to change it if there was some consensus on an alternative naming convention, though.

Reply:
I forgot that EVP_PKEY_CTX_new_id() also used the _id() suffix ... I guess I still prefer just _new(), but it is simply a matter of personal taste.
EVP_KDF_derive() makes sense to me.

Reply:
There are several ways of identifying an algorithm - short name, long name, ASN1_OBJECT, NID - and I imagined that there could be other `EVP_KDF_CTX_new_*()` variants in the future.
I was not sure if it should be `*_new_id()`, `*_new_nid()` or `*_new_bynid()`, etc., but there was an existing `*_new_id()` so I went with that.